src/devices/sound/lc82310.cpp

```diff
@@ -0,0 +1,296 @@
+{
+	stream = stream_alloc(0, 2, 44100);
+	mp3dec = std::make_unique<mp3_audio>(static_cast<const uint8_t *>(&mp3data[0]));
+
+	save_item(NAME(mp3data));
+	save_item(NAME(samples));
+	save_item(NAME(m_mp3data_count));
+	save_item(NAME(m_sample_count));
+	save_item(NAME(m_samples_idx));
+	save_item(NAME(m_frame_channels));
+	save_item(NAME(m_output_gain));
+
+	save_item(NAME(m_csctl));
+	save_item(NAME(m_ckctl));
+	save_item(NAME(m_dictl));
+	save_item(NAME(m_doctl));
+	save_item(NAME(m_ctl_bits));
+	save_item(NAME(m_ctl_byte));
+	save_item(NAME(m_ctl_out_byte));
+}
+
+void lc82310_device::device_reset()
+{
+	std::fill(std::begin(m_output_gain), std::end(m_output_gain), 0);
+
+	m_csctl = 0;
+	m_ckctl = 0;
+	m_dictl = 0;
+	m_doctl = 0;
+	m_ctl_bits = 0;
+	m_ctl_byte = 0;
+	m_ctl_out_byte = 0;
+	m_ctl_state = ACCEPTING_CMD;
+
+	reset_playback();
+}
+
+void lc82310_device::reset_playback()
+{
+	std::fill(mp3data.begin(), mp3data.end(), 0);
+	std::fill(samples.begin(), samples.end(), 0);
+
+	mp3dec->clear();
+	m_mp3data_count = 0;
+	m_sample_count = 0;
+	m_samples_idx = 0;
+	m_frame_channels = 2;
+}
+
+WRITE_LINE_MEMBER(lc82310_device::zcsctl_w)
+{
+	m_csctl = state;
+}
+
+WRITE_LINE_MEMBER(lc82310_device::ckctl_w)
+{
+	if (m_csctl == 0 && m_ckctl == 0 && state == 1)
+	{
+		m_ctl_byte |= m_dictl << m_ctl_bits;
+		m_ctl_bits++;
+
+		if (m_ctl_bits > 7)
+		{
+			if (m_ctl_state == ACCEPTING_CMD)
+			{
+				// Expected to be able to read the return value while sending the second byte
+				// Everything in the 0x80 range of commands seems to respond with a value
+				// 0x80, 0x81, 0x82 return separate 8-bit values
+				// 0x83 returns an 8-bit error status, bits 0 and 1 are used to signal errors, any non-0 bit is considered an error
+				// 0x84 and 0x85 are used together to form a 16-bit value
+				m_ctl_out_byte = 0;
+
+				m_ctl_cmd = m_ctl_byte;
+				m_ctl_state = ACCEPTING_PARAM;
+			}
+			else if (m_ctl_state == ACCEPTING_PARAM)
+			{
+				handle_command(m_ctl_cmd, m_ctl_byte);
+				m_ctl_state = ACCEPTING_CMD;
+			}
+
+			m_ctl_byte = 0;
+			m_ctl_bits = 0;
+		}
+
+		m_doctl = m_ctl_out_byte & 1;
+		m_ctl_out_byte >>= 1;
+	}
+
+	m_ckctl = state;
+}
+
+WRITE_LINE_MEMBER(lc82310_device::dictl_w)
+{
+	m_dictl = state;
+}
+
+READ_LINE_MEMBER(lc82310_device::doctl_r)
+{
+	return m_doctl;
+}
+
+READ_LINE_MEMBER(lc82310_device::demand_r)
+{
+	return m_mp3data_count < mp3data.size();
+}
+
+void lc82310_device::dimpg_w(uint8_t data)
+{
+	if (m_mp3data_count >= mp3data.size())
+	{
+		// Drop a byte if the buffer is full and it's still trying to send data
+		std::copy(mp3data.begin() + 1, mp3data.end(), mp3data.begin());
+		m_mp3data_count--;
+	}
+
+	mp3data[m_mp3data_count++] = data;
+}
+
+void lc82310_device::handle_command(uint8_t cmd, uint8_t param)
+{
+	if (cmd == CMD_UNK13_VOL || cmd == CMD_UNK15_VOL)
+	{
+		// These are calculated based on the configurable values in-game vs what is sent to the MP3 decoder
+		constexpr float gain_table[] = {
```

Comment:
If there's some simple method of calculating this that I'm missing, please let me know. I gave up and just made a table with all known values and interpolated values for the unknown ones.

Reply:
There may not be, it looks like stuff that could be in a rom.  Are you sure of the values for entries 23/24/25?  The shape there is a little weird.  It can be very roughly approximated with 10**(-x/32) or so.  Which are the known values in the table?

Reply:
The values I used came from Nice Tsukkomi, which has a function to look up the in-game volume values of 0-31.

This is the mapping from Nice Tsukkomi. The key on the left side is the in-game volume setting, and the value on the right side is what is sent to the MP3 decoder. So a key of 0 is for silent which corresponds to 0x4c/76 on the MP3 decoder, and the key of 31 is for max volume which becomes 0 on the MP3 decoder side. Everything here is exactly as is in the table so none of the interoplated values are included. So, the values for 24 and 25 are correct but 23 was calculated as the volume between keys 8 and 9.

```
0: 76
1: 42
2: 36
3: 32
4: 30
5: 28
6: 26
7: 25
8: 24
9: 22
10: 21
11: 20
12: 19
13: 18
14: 17
15: 16
16: 15
17: 14
18: 13
19: 12
20: 11
21: 10
22: 9
23: 8
24: 7
25: 6
26: 5
27: 4
28: 3
29: 2
30: 1
31: 0
```

Reply:
Oh I see.  And there's no reason to think the 0-31 is linear, expecially since they're hitting a resolution issue anyway.  So without any docs for the mp3 chip itself or hardware experiments there's no way to be sure of anything.  Oh well.